### PR TITLE
aruha-399 compatibility mode

### PR DESCRIPTION
**Pull request by rcillo**
Documentation for the three compatibility modes. It focuses on the
guarantees provided by each mode so that users know what to expected
from event types using each one of them.

### CODE REVIEW

File: docs/using/event-types.md [hunk 1]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
 `{ "\additionalProperties\": true }` to allow arbitrary JSON, but will usually
 have a more specific definition :)
 
+### Compatibility modes
+
+Event type schemas evolve through time and different data integration
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `time and` -> `time, and`

File: docs/using/event-types.md [hunk 2]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+patterns require different guarantees when it comes to what changes
+should or should not be allowed.
+
+To accomodate these different needs Nakadi provides three distinct
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `accomodate` -> `accommodate`

Comment (lmontrieux) [reply docs/using/event-types.md]: You can also remove `distinct` (only if you want to), the sentence still works without it

File: docs/using/event-types.md [hunk 3]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+
+Clients using older schemas can safely read events validated by newer
+schemas. It's not safe for a client using the latest schema to read
+events generated by older schemas, in the sense that there might be
+type inconsistencies due to additionalProperties.
+
+The complete set of json-schema is supported, including
+additionalProperties. Consumers willing to read events in a safe
+manner should ignore fields that are not declared in the
+schema. Making assumptions about non declared properties might lead to
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `non declared` -> `undeclared`?

File: docs/using/event-types.md [hunk 4]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+breaking changes.
+
+#### compatible
+
+This is the most strict mode. It guarantees that events published
+under this mode can be safely read using both newer and older schema
+versions. E.g. it's both forward compatible and backwards
```
Comment (lmontrieux) [inline docs/using/event-types.md]: I think you mean i.e., not e.g.?

File: docs/using/event-types.md [hunk 5]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+compatible. In other words:
+
+1. Clients using old schemas can safely read events validated by newer
+   schemas.
+2. Clients using the newest schema can safely read events validated by
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `newest` -> `newer`, if I'm not mistaken

File: docs/using/event-types.md [hunk 6]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+   older schemas.
+
+In order to achieve that, only a subset of json schema will be allowed
```
Comment (lmontrieux) [inline docs/using/event-types.md]: Earlier you wrote `json-schema`, not `json schema`

File: docs/using/event-types.md [hunk 7]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+to make sure no accidental incompatibilities or ambiguities will be
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `will be` -> `are`

File: docs/using/event-types.md [hunk 8]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+introduced. Only compatible changes aregoing to be allowed when
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `aregoing` -> `are going`

Comment (lmontrieux) [reply docs/using/event-types.md]: In fact, `aregoing to be` -> `are`

File: docs/using/event-types.md [hunk 9]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+changing the schema. This mode will be required to have data stored
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `will be` -> `is` (unless we want to change this again when it becomes true)

File: docs/using/event-types.md [hunk 10]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+and processed by the data lake.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `to have data stored and processed by the data lake` -> `for the data lake to store and process data` (removing passive form)

File: docs/using/event-types.md [hunk 11]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+
+#### none
+
+Schemas can be changed arbitrarily. Nakadi will automatically manage
+the version field so that if an incompatible change happens (MAJOR
+version bump) consumers are going to be aware of it.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `are going to be` -> `are`

File: docs/using/event-types.md [hunk 12]
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
...
+3. addition of optional attributes.
+
+All other changes are considered incompatible.
+
+#### Strict validation
+
+When using **compatible** mode, validation of events accept only
+properties that are explicitly stated in the schema. It means that
+events containing attributes that cannot be mapped to a schema
+property are going to be rejected.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `are going to be` -> `will be`

**Comment by rcillo** [comment]
@lmontrieux thank you for the previous review. I applied the changes when appropriate but changed a lot the text. It wasn't properly providing all the context, behaviour, limitations and use case for each mode.

I hope it's clearer now.

File: docs/using/event-types.md [hunk 13]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
 `{ "\additionalProperties\": true }` to allow arbitrary JSON, but will usually
 have a more specific definition :)
 
+### Compatibility modes
+
+Compatibility modes are used to control schema changes. Each mode
+solves a specific problem and thus present different constraints. It
```
Comment (lmontrieux) [inline docs/using/event-types.md]: present -> presents

File: docs/using/event-types.md [hunk 14]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+ranges from being very constraining, allowing just some minor changes
```
Comment (lmontrieux) [inline docs/using/event-types.md]: comma (,) after 'changes'

File: docs/using/event-types.md [hunk 15]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+to allowing everything.
+
+By default, Nakadi uses the forward mode, which is further explained
+bellow, together with the other two modes: compatible and none.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `bellow` -> `below`

Comment (lmontrieux) [reply docs/using/event-types.md]: `the other two modes` -> `the two other modes` (I think)

File: docs/using/event-types.md [hunk 16]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+
+#### Forward
+
+This mode has been designed to allow event type owners to change their
+schemas without breaking existing consumers.
+
+Its called forward because using older schemas it's safe to read
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `Its` -> `It's` (or better, `It is`)

Comment (lmontrieux) [reply docs/using/event-types.md]: comma after `older schemas`

File: docs/using/event-types.md [hunk 17]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+events generated by newer schemas. On the other hand, it's not
+backward, in the sense that trying to read an older event using a
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `backward` -> `backward compatible`, if I'm not mistaken?

File: docs/using/event-types.md [hunk 18]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+newer schema is not safe.
+
+So if the consumer has been developed with schema version `1.5.0` in
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `So` -> `For example,`

File: docs/using/event-types.md [hunk 19]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+mind, reading events from `1.5.0` on is safe, for example, `1.6.0`. But
+reading event generated by older schemas, like `1.4.0` is not safe.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: comma after `1.4.0`

File: docs/using/event-types.md [hunk 20]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+
+1. changes to meta attributes: titles, descriptions and examples.
+2. addition of new attributes.
+3. marking attributes as required.
+
+Consumers reading events from forward compatible event types SHOULD
+ignore event attributes not declared in the event type schema.
+
+This mode is default because it solves the most common requirement:
+guarantee stability of deployed consumer applications when schema
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `guarantee stability` -> `guarantee the stability`
`consumer` -> `consumers`
`when schema` -> `when the schema`

File: docs/using/event-types.md [hunk 21]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+
+This is the most strict mode. It's both forward compatible and
+backward compatible. In practical terms:
+
+1. Consuemrs using older schemas can safely read events validated by
+   newer schemas.
+2. Clients using newer schema can safely read events validated by
+   older schemas.
+
+It's useful for long term data storage and analysis. It's guarantees
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `It's guarantees` -> `It guarantees`

File: docs/using/event-types.md [hunk 22]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+the highest quality data for business intelegence.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `intelegence` -> `intelligence`

File: docs/using/event-types.md [hunk 23]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+
+Only a subset of json schema is allowed to make sure no accidental
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `allowed to` -> `allowed, to`
`sure no` -> `sure that no`

File: docs/using/event-types.md [hunk 24]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+incompatibilities or ambiguities will be introduced. This mode is
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `will be` -> `are`

File: docs/using/event-types.md [hunk 25]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+required to have data stored and processed by the data lake.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `to have data` -> `for data to be`
OR
change the whole line to `required for the data lake to store and process the data.`

File: docs/using/event-types.md [hunk 26]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+This mode is default because it solves the most common requirement:
+guarantee stability of deployed consumer applications when schema
+changes. It's crucial for the robustness of event driven applications.
+
+#### Compatible
+
+This is the most strict mode. It's both forward compatible and
+backward compatible. In practical terms:
+
+1. Consuemrs using older schemas can safely read events validated by
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `Consuemrs` -> `Consumers`

File: docs/using/event-types.md [hunk 27]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+newer schema is not safe.
+
+So if the consumer has been developed with schema version `1.5.0` in
+mind, reading events from `1.5.0` on is safe, for example, `1.6.0`. But
+reading event generated by older schemas, like `1.4.0` is not safe.
+
+Supported changes:
+
+1. changes to meta attributes: titles, descriptions and examples.
+2. addition of new attributes.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: both optional and required, right? Maybe this should be explicit, as in the Compatible you say that only new optional attributes are allowed

File: docs/using/event-types.md [hunk 28]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+2. addition of new optional attributes.
+
+Unsupported json-schema attributes:
+
+1. additionalProperties
+2. additionalItems
+3. not
+4. patternProperties
+
+Removing the support to these features is necessary to avoid the
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `support to` -> `support for`

File: docs/using/event-types.md [hunk 29]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+introduction of incompatible changes. Since additionalProperties is
+not supported, it's value is defaulted to false: all events with
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `it's value is defaulted to false` -> `its value is false by default`

File: docs/using/event-types.md [hunk 30]
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
...
+undeclared attributes rejected.
```
Comment (lmontrieux) [inline docs/using/event-types.md]: `attributes rejected` -> `attributes will be rejected`

**Review by lmontrieux — COMMENTED** [review]
Look very clear and detailed to me. I mostly made comments about grammar and typos.

**Comment by rcillo** [comment]
👍 made all the corrections. I realised my grammar checker was disabled 😅

**Comment by lmontrieux** [comment]
👍

**Pull request merged**